### ISSUE TITLE
Add support to store the domain events in an "outbox" table

### ISSUE THREAD

**Issue by dsantang** [dsantang/domain-events-doctrine#5]
## Feature: store outbox formatted domain events in the outbox table

In order to avoid "dual writes", first on the database and then on Kafka, we want to make sure that our applications can rely on an implementation of the "outbox" pattern to safely publish those events.
Currently, in order to publish relevant domain events to Kafka, we're relying on top of 2 libraries: `dsantang\domain-events` and `dsantang\domain-events-doctrine`.
`dsantang\domain-events` enables raising events from our aggregates, while instead `dsantang\domain-events-doctrine` makes sure it's possible to dispatch those events via a Symfony's event dispatcher on Doctrine's `postFlush` event.

### requirements:

- It should be possible to store in an outbox table, an ordered list of events, that respects the format stated by: https://issues.jboss.org/browse/DBZ-1169.
  Those events will later be published to a queue/stream by a CDC system like Debezium.
- The events should be stored in the same order they were published (or make sure that the CDC system is able to publish them in the correct order by any mean).
- It should be possible, for a Client using the library, to modify (for example by enriching) the event's payload data before storing it in the outbox table.

### Technical constraints:

- The outbox logic/data should never leak to the `DomainEvent` itself.

### Considerations:

It is possible to hook in into Doctrine's lifecycle via Doctrine's events manager, and that's what is currently being used by `dsantang\domain-events-doctrine` in order to dispatch the domain events.
We also know that it is possible to add entities to an "ongoing" transaction by hooking into Doctrine's `onFlush` event (ref: https://stackoverflow.com/questions/15092965/create-entity-on-entity-flush).
This enables us to create "outbox" entries based on the domain events risen by the application, and safely store them in the same DB transaction.

**Comment by fabiotc** [comment]
I just opened a PR https://github.com/dsantang/domain-events-doctrine/pull/6 (still work in progress) with the first implementation attempt.
There are still some missing implementations there that I need your help to know how to proceed.

#### OutboxEvent interface:
```PHP
interface OutboxEvent extends DomainEvent
{
    public function getRoute() : string;
    public function getAggregateId() : UuidInterface;
    public function getAggregateType() : string;
    public function getPayloadType() : string;
}
```
As you can see this implementation can give a lot of power to the events to inject all the needed information, but on the other hand, it's too verbose, requiring a lot of code lines.
So please give some suggestions if I need to change/remove something.
And I think it's better to move this class to domains-event library, but I'm not completely sure about it.

#### OutboxTransformer
There's a class `(OutboxTransformer)` with the responsibility to transform a Domain Event (in this case an Outbox Event) in an Outbox entity (to be stored).
So it will be easy to change the behavior of the transformation (please note that the implementation of this class is still incomplete in the PR).

So basically we need to decide:

From OutboxEvent interface? (like here: https://github.com/dsantang/domain-events-doctrine/blob/feature/outbox-workflow/tests/RandomOutboxEvent.php):
- `messageRoute`
- `messageType`
- `aggregateId`
- `aggregateType`
- `payloadType`

And the other fields:
- `messageType` - I'm getting from the name of the EventClass
- `payload` - Not implemented yet. Should it be a fully EventClass converted to JSON?
